### PR TITLE
floating label gap patch

### PR DESCRIPTION
**Pull request by lastmeta**
Parameterize a hardcoded number: the gap between the label on a TextField and the content of the TextField.

*List which issues are fixed by this PR. You must list at least one issue.*
https://github.com/flutter/flutter/issues/130030

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/material/input_decorator.dart
Comment (goderbauer) [inline packages/flutter/lib/src/material/input_decorator.dart:3940]: This macro doesn't exist, to define it you will have to use `{@template flutter.material.inputDecoration.floatingLabelGap} ... {@endtemplate}` somewhere else in a doc (see other examples of `{@template` in our repo)

File: packages/flutter/lib/src/material/input_decorator.dart
Comment (goderbauer) [inline packages/flutter/lib/src/material/input_decorator.dart:2908]: Docs should be grammatical correct sentences, so this one should end with a `.`.

**Review by goderbauer — COMMENTED** [review]
You will also have to add a test to this (see message by the bot posted above).

I'll leave it to the InputDecorator experts to review this further. cc @justinmc

File: packages/flutter/lib/src/material/input_decorator.dart
Comment (lastmeta) [reply packages/flutter/lib/src/material/input_decorator.dart]: done

File: packages/flutter/lib/src/material/input_decorator.dart
Comment (justinmc) [inline packages/flutter/lib/src/material/input_decorator.dart:2408]: This means that your floatingLabelGap will change with the text scale factor.  So if a users sets their system font size to larger than normal in their device settings, then the gap will increase.  Do you think that's desired?

If we change it so that text scale factor doesn't affect this gap, then likely some visual diff tests will be broken.

Comment (justinmc) [reply packages/flutter/lib/src/material/input_decorator.dart:2408]: Also could the `4.0` be set as the default value for the floatingLabelGap parameter?

**Review by justinmc — COMMENTED** [review]
Left a comment about textScaleFactor.  Also, this will need a test please!

I was also somewhat concerned about making InputDecorator's layout more complex because it already is a source of bugs.  But I guess this change doesn't increase complexity too much.

File: packages/flutter/lib/src/material/input_decorator.dart
Comment (lastmeta) [reply packages/flutter/lib/src/material/input_decorator.dart:2408]: > if a users sets their system font size to larger than normal in their device settings, then the gap will increase. Do you think that's desired?

This is what happens now. My change reflex as minimal alternation as possible.

Comment (lastmeta) [reply packages/flutter/lib/src/material/input_decorator.dart:2408]: > Also could the `4.0` be set as the default value for the floatingLabelGap parameter?

I think it could be, but it was simpler to code this way. I don't remember the issue I ran into when I tried it that way first but I switched it to this because I hit a snag.

**Comment by lastmeta** [comment]
Thanks so much for reviewing this slight alteration! I want to say, I don't do much opensource stuff, but since I already solved this issue for my project (by overriding these widgets manually in the same way I showed in the pull request), I thought it might be nice to share it with the community. I saw a few other's who have had the same challenge.

> I was also somewhat concerned about making InputDecorator's layout more complex because it already is a source of bugs. But I guess this change doesn't increase complexity too much.

Yes, as you can see I'm not making the layout more complex at all. I'm parameterizing something that might just as well have been a parameter in the first place. This code has zero affect on anything, as the original structure is untouched. I merely optionally parameterized a hard coded number. I don't even move that hard coded number away from where it originally was in the code.

Now, by parameterizing this I've opened the door for the 4.0 value in gap equation to be something other than the constant 4.0 which could possibly cause problems elsewhere, but I can't begin to imagine where or why. I don't think anything _should_ rely on that.

> Also, this will need a test please!

So I'm not sure I _can_ write a test that discovers any substantial problem this change may lead to.

I know this might be annoying but I'm not all that motivated to dig into the test suite when I've already solved this issue for my project, and I know its potential problems are limited to bizarre edge cases of radically manipulating that parameter, (if there are any problems whatsoever; I already know it can withstand negative values without error in my situation).

And those problems are not something I can easily test for. Maybe, being familiar with the test suite and the other internals of flutter, you or someone else can more easily?

I don't want to be difficult, maybe if you can outline the test you think would be valuable I could write it.

**Comment by justinmc** [comment]
Since you've solved this for yourself already, maybe we should wait for feedback from anyone else experiencing this problem before merging instead of spending any more of your time getting this ready.  I'll post comments on the issues.  If this doesn't attract interest we can close it for the time being.

In case we come back to this, I would test this by adding something to input_decorator_test kind of like [this test](https://github.com/flutter/flutter/blob/9a726e8542681326d91d0d750d3f268941603d0c/packages/flutter/test/material/input_decorator_test.dart#L166).  I'd pass in floatingLabelGap and expect that the location of the label has changed.

**Comment by lastmeta** [comment]
> Since you've solved this for yourself already...

Yes, I've solved it by copying the input decorator and form field widgets out of material and then changed them to add this very simple parameterization. But that's not a permanent solution - I have to manually update this code as flutter updates.

I've added a test where you suggested.

I'm not sure the problem that is being solved is clear, but it is really really simple, so I'd like to reiterate it with images here if I could:

These designs are impossible to recreate:
image
image
image

They're impossible for only one reason: There is no way to shorten this distance (it's hard coded).

image
image
image

The hard coded value looks something like this:

image

Designs that require the label to be "inside" the border require the user to make a very large custom, non-responsive border. We should, instead, parameterize the "gap" value so we can move the label closer to the text, and allow for designs that have the label inside the border.

Simple, safe, and easy. 

Here's a list of people running into this problem, or it's variant, apart from previously referenced issues:

https://stackoverflow.com/questions/71121025/how-to-change-the-labeltext-position-a-little-down
https://stackoverflow.com/questions/68539705/flutter-how-to-make-text-form-field-with-floating-label-intro-container
https://stackoverflow.com/questions/71621132/how-to-make-the-floating-label-stay-within-the-border
https://stackoverflow.com/questions/69071964/flutter-input-text-field-with-custom-label-position

I think it's ready to merge.

File: packages/flutter/lib/src/material/input_decorator.dart
Comment (justinmc) [inline packages/flutter/lib/src/material/input_decorator.dart:2408]: You have conflicts in this file, it might be due to this recent change from textScaleFactorOf to textScalerOf.  That's also causing an analyzer failure.

File: packages/flutter/test/material/input_decorator_test.dart
Comment (justinmc) [inline packages/flutter/test/material/input_decorator_test.dart:391]: Nit: Period at the end here and elsewhere.

File: packages/flutter/test/material/input_decorator_test.dart
Comment (justinmc) [inline packages/flutter/test/material/input_decorator_test.dart:436]: I think some text got cut off at the end of this comment.

File: packages/flutter/test/material/input_decorator_test.dart
Comment (justinmc) [inline packages/flutter/test/material/input_decorator_test.dart:453]: This comment is also cut off.

File: packages/flutter/test/material/input_decorator_test.dart
Comment (justinmc) [inline packages/flutter/test/material/input_decorator_test.dart:512]: "it's" => "its"

File: packages/flutter/test/material/input_decorator_test.dart
Comment (justinmc) [inline packages/flutter/test/material/input_decorator_test.dart:626]: Can you split this test up at each `pumpWidget`?  That way we'll get more information when there's a failure.

**Review by justinmc — COMMENTED** [review]
Alright let's do it.  Thanks for putting in the effort to figure out the test.  Just some small comments here.

**Comment by lastmeta** [comment]
> Alright let's do it. Thanks for putting in the effort to figure out the test. Just some small comments here.

cool, I'll work on those conflicts today

File: packages/flutter/test/material/input_decorator_test.dart
Comment (lastmeta) [reply packages/flutter/test/material/input_decorator_test.dart:516]: done.

File: packages/flutter/test/material/input_decorator_test.dart
Comment (lastmeta) [reply packages/flutter/test/material/input_decorator_test.dart:457]: done.

File: packages/flutter/test/material/input_decorator_test.dart
Comment (lastmeta) [reply packages/flutter/test/material/input_decorator_test.dart:395]: done.

File: packages/flutter/lib/src/material/input_decorator.dart
Comment (lastmeta) [reply packages/flutter/lib/src/material/input_decorator.dart:2409]: resolved.

**Comment by justinmc** [comment]
@lastmeta Do you still have plans to finish this?  No rush, just checking in.

**Comment by lastmeta** [comment]
> @lastmeta Do you still have plans to finish this? No rush, just checking in.

I already did everything you mentioned. I'm not sure what the issue is at this point.

**Comment by justinmc** [comment]
You have a bunch of test failures.  In the code it looks like you forgot a merge conflict `<<<<<<< HEAD` so maybe that's the only problem.

**Comment by lastmeta** [comment]
>

**Comment by lastmeta** [comment]
> You have a bunch of test failures. In the code it looks like you forgot a merge conflict `<<<<<<< HEAD` so maybe that's the only problem.

Thanks, don't know how I missed that one, most of the tests are passing now, but there's still some issues.

**Comment by justinmc** [comment]
Hmm looks like there are legitimate failures in input_decorator_test.dart, including at least one of the tests that you added in this PR.

File: packages/flutter/test/material/input_decorator_test.dart
Comment (justinmc) [inline packages/flutter/test/material/input_decorator_test.dart:391]: Nit: Space after the `//`, here and below.

File: packages/flutter/test/material/input_decorator_test.dart
Comment (justinmc) [inline packages/flutter/test/material/input_decorator_test.dart:552]: Nit: "it's" => "its"

**Review by justinmc — APPROVED** [review]
LGTM with nits :+1:.  Thanks for fixing the tests!

File: packages/flutter/test/material/input_decorator_test.dart
Comment (lastmeta) [reply packages/flutter/test/material/input_decorator_test.dart]: done

File: packages/flutter/test/material/input_decorator_test.dart
Comment (lastmeta) [reply packages/flutter/test/material/input_decorator_test.dart]: done

**Comment by XilaiZhang** [comment]
(sorry for the notifications, but I am selecting this PR to test some Google Testing logic. this is the only open PR i can find at the moment, that is created by an external contributor, and also has got a googler approval)

**Comment by lastmeta** [comment]
> (sorry for the notifications, but I am selecting this PR to test some Google Testing logic. this is the only open PR i can find at the moment, that is created by an external contributor, and also has got a googler approval)

Is there anything I need to do to progress this? (other than attempting to keep it up to do date with flutter, I see there are some new conflicts to resolve...)

**Comment by XilaiZhang** [comment]
yeah for the merge conflicts, you can have a Googler to help investigate if its okay to override, or wait for flutter rolls to catch up. if its low risk, you can skip the wait and override this check

**Comment by lastmeta** [comment]
> if its low risk, you can skip the wait and override this check

It is low risk (it's only parameterizing a hardcoded used in one place). But I don't have the ability to override anything because I'm the external contributor.

Thanks for your attention to this. I'm willing to wait, but if anyone wants to take a look at it, verify it's really simple and safe, and push it along I'd appreciate it!

**Comment by lastmeta** [comment]
@goderbauer Anything you need from me?

**Comment by LongCatIsLooong** [comment]
It looks like increasing the gap actually increases the height of the field by roughly 2x that value. This is when I set the gap to 200. The red square at the bottom has a height of 200 px.

https://github.com/flutter/flutter/assets/31859944/84cd712c-c4ae-4ad9-ad7d-c2cc21672ea1

File: packages/flutter/lib/src/material/input_decorator.dart
Comment (lastmeta) [reply packages/flutter/lib/src/material/input_decorator.dart:2926]: sure.

**Comment by lastmeta** [comment]
> It looks like increasing the gap actually increases the height of the field by roughly 2x that value. This is when I set the gap to 200. The red square at the bottom has a height of 200 px.

That intended. It is how flutter works currently. Here is the essence of my alteration:

FROM:
```
(4 + 0.75 * labelStyle.fontSize!) * MediaQuery.textScaleFactorOf(context);
```
TO:
```
(decoration.floatingLabelGap + 0.75 * labelStyle.fontSize!) * MediaQuery.textScaleFactorOf(context);
```

It is my intent to avoid changing any functionality, whatsoever. This alteration ***merely parameterizes a hardcoded value.***

**Comment by LongCatIsLooong** [comment]
The documentation says the value represents the gap between the inline element, so I'd expect the gap to be removed when there's no inline elements but in this implementation that doesn't seem to be the case? IMO it probably makes more sense to interpolate the gap value during the floating label animation.

the documentation says "Actual calculation is:  `(floatingLabelGap + 0.75 * labelStyle.fontSize!) * MediaQuery.textScaleFactorOf(context);`", but the value seems to be used as the height of the floating label not the gap right?

**Comment by lastmeta** [comment]
> but the value seems to be used as the height of the floating label not the gap right?

Yeah I guess. I don't know technically, and my assumption was not that technically - I assumed the height of the entire input field was determined in part by the maximum gap + the size of the input text area. Is there a difference between that and your observation that his defines that maximum height of the label area? seems like a distinction without a difference to me.

Do you guys not think this is an improvement? I wanted control over the height of the animated label above the text input and I don't have control over that until I parameterized this (or some value) that should have already been a parameter. Now I have control over that and can implement the design of the textfield.

> The documentation says the value represents the gap between the inline element, so I'd expect the gap to be removed when there's no inline elements but in this implementation that doesn't seem to be the case? IMO it probably makes more sense to interpolate the gap value during the floating label animation.

As I said earlier I don't want to change any code I don't have to. So if you think the documentation should change to say it's the height, as opposed to the gap, I'm happy to do that but I don't think I should be in there mucking around with any code not directly related to parameterizing a hard coded value that should have been parameterized in the first place.

**Comment by LongCatIsLooong** [comment]
Looking at the https://github.com/flutter/flutter/issues/130030 it seems you'd like to reduce the gap to 0? From the material 3 spec there doesn't seem to be a gap between the input text and the floating label in a populated text field: https://m3.material.io/components/text-fields/specs#0d36c3fe-7948-4ec2-ab8a-4fe39cca19cc, the current height calculation `floatingLabelHeight = (4.0 + 0.75 * labelStyle.fontSize!) * MediaQuery.textScalerOf(context).textScaleFactor;` seems to be a different from the spec and it doesn't respect any other typographic settings specified in `labelStyle` other than `fontSize` (e.g., `TextStyle.height` is ignored`). I'll try to create a PR that removes that logic and uses the real height of the label measured from the layout pass.

> Do you guys not think this is an improvement? I wanted control over the height of the animated label above the text input and I didn't have control over that until I parameterized this. Others have wanted the same thing and come up with all kinds of more complicated requests for achieving essentially that (see related issues for more on that).

Without clear use cases that support the new public API, it's hard to tell exactly what the API is supposed to do and how it should behave in different text field states (and as a result it will be hard to document). My apologies for the frustration but I think if the API is confusing (people will have to look at the source code or figure out the right value to set through trial and error) then it will probably cost more people more dev hours to understand or maintain. If the goal is to reduce the vertical gap between the input text and the floating label I think the change proposed above should do what you want (and you should be able to change the line height of the label via `TextStyle.height` to increase/reduce the gap).

**Comment by lastmeta** [comment]
> the current height calculation `floatingLabelHeight = (4.0 + 0.75 * labelStyle.fontSize!) * MediaQuery.textScalerOf(context).textScaleFactor;` seems to be a little different from the spec and it doesn't respect any other typographic settings specified in `labelStyle` other than `fontSize` (e.g., `TextStyle.height` is ignored`). I'll try to create a PR that removes that logic and uses the real height of the label measured from the layout pass.
...If the goal is to reduce the vertical gap between the input text and the floating label I think the change proposed above should do what you want (and you should be able to change the line height of the label via `TextStyle.height` to increase/reduce the gap).

If there's an even better solution more inline with the Material specs, that's fine with me, I just want control over that distance, thank you!

**Comment by LongCatIsLooong** [comment]
FYI there's a plan to rewrite the input decorator which hopefully is going to address this issue so I'm going to close this for now (see the linked issue for a workaround). @HansMuller is there an issue for the input decorator rewriting?

**Pull request closed without merging**
